SpringParking-parkingBranch/src/main/java/pl/springDemo/demo/userPayment/UserPaymentServiceImpl.java

```diff
@@ -0,0 +1,67 @@
+package pl.springDemo.demo.userPayment;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import pl.springDemo.demo.user.UserRepo;
+import pl.springDemo.demo.user.UserType;
+
+import java.math.BigDecimal;
+import java.time.Duration;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+
+@Service
+public class UserPaymentServiceImpl implements UserPaymentService {
+
+    private final UserPaymentRepo userPaymentRepo;
+    private final UserRepo userRepo;
+    private final AmountCounter amountCounter;
+
+    @Autowired
+    public UserPaymentServiceImpl(UserPaymentRepo userPaymentRepo, UserRepo userRepo, AmountCounter amountCounter) {
+        this.userPaymentRepo = userPaymentRepo;
+        this.userRepo = userRepo;
+        this.amountCounter = amountCounter;
+    }
+
+    @Override
+    public void createStartDate(long userPaymentId) {
+        LocalDateTime dateFrom = LocalDateTime.now();
+        UserPayment userPayment = userPaymentRepo.findById(userPaymentId).get();
+        userPayment.setDateFrom(dateFrom);
+        userPaymentRepo.save(userPayment);
+    }
+
+    @Override
+    public void createEndDate(long userPaymentId) {
+        LocalDateTime dateTo = LocalDateTime.now();
+        UserPayment userPayment = userPaymentRepo.findById(userPaymentId).get();
+        userPayment.setDateTo(dateTo);
+        userPaymentRepo.save(userPayment);
+    }
+
+    @Override
+    public boolean hasUserTicket(long userId) {
+        return userPaymentRepo.findUserWithStartDate().contains(userId);
+    }
+
+    @Override
+    public double calculatePayment(long userId) {
+        UserType userType = userRepo.findUserTypeById(userId);
+        LocalDateTime start = userPaymentRepo.findDateFromById(userId);
+        LocalDateTime end = userPaymentRepo.findDateToById(userId);
+
+        long hours = Duration.between(start,end).toHours();
```

Comment:
Jeżeli kierowca zaparkuje na 90 minut to jaka będzie opłata? Za 1 czy 2 godziny? Dla takich przypadków przydają się testy.

Reply:
Zdecydowanie mogłam dopisać więcej testów, czy w ogóle test tej klasy. Zgadzam się, to mój błąd. Nie było określone pokrycie kodu testami, jednak oczywiście oczyściłoby to jeszcze kod.